bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+        rho: VectorType,
+        stepsize: float,
+        steps: int,
+    ) -> tuple[VectorType, VectorType]:
+        """Propose new sample (theta_prop, rho_prop) with leapfrog integrator.
+
+        Discretize and solve Hamilton's equations using the leapfrog integrator to
+        deterministically propose a new sample (theta_prop, rho_prop) from the current
+        sample (theta, rho).
+
+        Copy theta because numpy's += operator mutates the original array instead of
+        creating a new one.
+
+        Initialize rho_mid by going backwards half a step so that the first full-step
+        inside the loop brings rho_mid up to +1/2 steps.
+
+        Args:
+            theta: position
+            rho: momentum
+            stepsize: stepsize in each leapfrog step
+            steps: number of leapfrog steps
+
+        Returns:
+            tuple[VectorType, VectorType]: proposed sample
+        """
+        theta = np.array(theta, copy=True)
+        _, grad = self._model.log_density_gradient(theta)
+        rho_mid = rho - 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        for _ in range(steps):
+            rho_mid += stepsize * np.multiply(self._metric, grad).squeeze()
+            theta += stepsize * rho_mid
+            _, grad = self._model.log_density_gradient(theta)
+        rho = rho_mid + 0.5 * stepsize * np.multiply(self._metric, grad).squeeze()
+        return (theta, rho)
+
+    def _prob_of_delayed_rejection(self, hastings: float) -> float:
+        """Log probability of making another proposal upon rejection.
+
+        To reduce average cost per iteration for DRHMC, make the delayed rejections
+        probabilistic, such that a subsequent proposal is not mandatory upon rejection.
+
+        Probabilistic delayed rejection proposes a new sample with some probability
+        dependent on where we are in the distribution. If a proposal has low acceptance
+        probability and was rejected, propose another sample; if a proposal has high
+        acceptance probablity but was rejected by chance, do not propose another
+        sample. More details found in section 3.2 of Modi et al.
+
+        A heuristic proposal probability that achieve this is eqn. 31 in Modi et al:
+        the probability of making proposal k+1, after rejecting proposal k, is the
+        probability of rejecting proposal k. The proposal probability of making the
+        first proposal, however, is always 1.
+
+        This proposal probability modifies the proposal kernel, which alters detailed
+        balance. Thus the probability of accepting a proposal is multiplied by the
+        hasting term of the propopsed sample divided by the hastings term of the
+        current sample.
+
+        Args:
+            hastings: log probability of rejecting all previous proposals
+
+        Returns:
+            float: log probability of making another proposal
+        """
+        return hastings if self._probabilistic else 0.0
```

Comment:
Shouldn't this be else `-np.inf`?  It should have a zero probability if `self._prob == False`, which means `-np.inf` log probability.  As is, if `self._probabilistic = False`, you always return log(1) = 0.

Reply:
We seek the probability of attempting another proposal, denoted by $p$, and its equivalent on the log scale $lp := \log(p)$.

When `self._probabilistic = False`, we want $p=1$ such that we *always* retry another proposal. This implies that $lp = 0$.

Thus the function `_prob_delayed_rejection()` returns $0$ because it returns lp, the *log* probability of attempting another proposal upon rejection.

From your comment below, it seems it was unclear that `_prob_delayed_rejection()` returns $lp$ on the log scale. With this clarified, is this comment resolved?